[SAP] Netapp RPC for Qtree

hemna opened the pull request:
This patch adds the needed code to add a remote rpc call into a running netapp nfs_cmode driver.

Change-Id: Iffd0683a06d56fee0d9921a91dbe6ad37163bdd9

leust commented:
I squashed and added the changes from #266 into this PR.

hemna commented on cinder/volume/drivers/netapp/dataontap/client/client_cmode_rest.py:
```diff
@@ -759,8 +759,11 @@ def get_lun_sizes_by_volume(self, volume_name):
+        volume = self._get_volume_by_args(vol_name=vol_name)
+        dir_rel_to_vol = '/'.join(dir_path.split('/')[1:])
+        # Path requires "%2E" to represent "." and "%2F" to represent "/".
+        dir_rel_to_vol = dir_rel_to_vol.replace('.', '%2E').replace('/', '%2F')
```
is this going to work when we switch over to kvm where we aren't guaranteed to have a qtree ?  This seems to override the previous vol_name=dir_path with the vol_name=vol_name for all cases?

Scsabiii replied:
Yes, this works with or without qtree.
Netapp made a bad assumption, that we have only 1 level volume depth and without escaping the URL the function does not work. So it is not directly qtree related, but multi-level directory structure support.
Even with kvm, we might have to structure the data and not put everything to a single root dir

hemna commented on cinder/volume/drivers/netapp/remote.py:
```diff
@@ -0,0 +1,82 @@
+    def test(self, ctxt, host):
+        LOG.warning("CAlling Netapp Remote API::test %s", host)
+        cctxt = self._get_cctxt(host=host)
+        return cctxt.call(ctxt, 'test')
```
probably can remove the test call now?

Scsabiii replied:
yes, we can remove it

hemna commented on cinder/scheduler/filters/sap_fcd_filter.py:
```diff
@@ -108,6 +119,23 @@ def backend_passes(self, backend_state, filter_properties):
+            orig_shard_ = orig_shard.replace('-', '_')
+            LOG.debug("orig_shard_: %s", orig_shard_)
+            LOG.debug("orig_pool: %s", orig_pool)
+            LOG.debug("tpool_vol: %s", tpool_vol)
```
Do we still need all these debug logs?  Can we reduce them to a single log ?

Scsabiii replied:
Yes, we can simplify it to one log line

hemna commented on cinder/volume/drivers/vmware/fcd.py:
```diff
@@ -889,13 +890,29 @@ def _get_fcd_location(self, fcd_id, datastore_ref):
+            try:
+                mpath = remote_ds_info['mount_path'].split(':')[1]
+                lpath = self.volumeops._get_mount_path(
+                    local_ds_info.datastore)
```
can you add a comment here what the lpath is supposed to look like and why we are doing these splits and tests?

hemna commented on cinder/volume/drivers/vmware/volumeops.py:
```diff
@@ -2716,6 +2725,63 @@ def _get_task_detail(task_ref, session):
+    def migrate_unattached_qtree(self, fcd_loc, tgt_ds_mpath,
+                                 new_disk_path, netapp_api,
+                                 netapp_host, context):
+        src_mpath = self._get_mount_path(fcd_loc.ds_ref())
```
can you add a comment here describing what the src_mpath looks like and what we are extracting from it?

Scsabiii committed:
This patch adds the needed code to add a remote rpc call into a running netapp nfs_cmode driver.
Fast volume migration via moving files beetween qtree dirs of the same netapp volume
Fix multi-level directory support in nfs_cmode driver

Change-Id: Ic1b2a0fbd9be04b65317215e1edf5861aad3b731
